// simple-git/build/responses/FetchSummary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class FetchSummary {
    constructor(raw) {
        this.raw = raw;
        this.remote = null;
        this.branches = [];
        this.tags = [];
    }
    static parse(data) {
        const fetchSummary = new FetchSummary(data);
        String(data)
            .trim()
            .split('\n')
            .forEach(function (line) {
            const original = line.trim();
            PARSERS.some(function (parser) {
                const parsed = parser[0].exec(original);
                if (parsed) {
                    parser[1](fetchSummary, parsed.slice(1));
                    return true;
                }
            });
        });
        return fetchSummary;
    }
}
exports.FetchSummary = FetchSummary;
const PARSERS = [
    [
        /From (.+)$/, function (fetchSummary, matches) {
            fetchSummary.remote = matches[0];
        }
    ],
    [
        /\* \[new branch\]\s+(\S+)\s*\-> (.+)$/, function (fetchSummary, matches) {
            fetchSummary.branches.push({
                name: matches[0],
                tracking: matches[1]
            });
        }
    ],
    [
        /\* \[new tag\]\s+(\S+)\s*\-> (.+)$/, function (fetchSummary, matches) {
            fetchSummary.tags.push({
                name: matches[0],
                tracking: matches[1]
            });
        }
    ]
];
