// simple-git/build/responses/ListLogSummary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ListLogLine_1 = require("./ListLogLine");
/**
 * The ListLogSummary is returned as a response to getting `git().log()` or `git().stashList()`
 *
 * @constructor
 */
class ListLogSummary {
    constructor(all) {
        this.all = all;
        this.latest = all.length && all[0] || null;
        this.total = all.length;
    }
    parse(text, splitter, fields) {
        fields = fields || ['hash', 'date', 'message', 'author_name', 'author_email'];
        return new ListLogSummary(text
            .split(ListLogSummary.COMMIT_BOUNDARY + '\n')
            .map(function (item) {
            return item.replace(ListLogSummary.COMMIT_BOUNDARY, '');
        })
            .filter(Boolean)
            .map(function (item) {
            return new ListLogLine_1.ListLogLine(item.trim().split(splitter), fields);
        }));
    }
}
ListLogSummary.COMMIT_BOUNDARY = '------------------------ >8 ------------------------';
exports.ListLogSummary = ListLogSummary;
exports.COMMIT_BOUNDARY = ListLogSummary.COMMIT_BOUNDARY;
